openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
-  allow number of states to be enlarged to be an integral multiple of number of GPUs.
-* Finish AMOEBA support.
-* Can alchemically-modified System objects share unmodified Force objects to avoid overhead
-  of duplicating Forces that are not modified?
+* `AlchemicalFactory` uses fused elecrostatic and steric alchemical modifications.
 
 """
 
-#=============================================================================================
+# TODO
+# - Generalize treatment of nonbonded sterics/electrostatics intra-alchemical
+#   forces to support arbitrary mixing rules. Can we eliminate decoupling to something simpler?
+# - Add support for other GBSA models.
+# - Add functions for the automatic optimization of alchemical states?
+# - Can we store serialized form of Force objects so that we can save time in reconstituting
+#   Force objects when we make copies?  We can even manipulate the XML representation directly.
+# - Allow protocols to automatically be resized to arbitrary number of states, to
+#   allow number of states to be enlarged to be an integral multiple of number of GPUs.
+# - Finish AMOEBA support.
+# - Can alchemically-modified System objects share unmodified Force objects to avoid overhead
+#   of duplicating Forces that are not modified?
+
+
+# =============================================================================
 # GLOBAL IMPORTS
-#=============================================================================================
+# =============================================================================
 
-import numpy as np
 import copy
-import time
-import itertools
+import inspect
+import logging
+import collections
 
-import simtk.openmm as openmm
-import simtk.unit as unit
+import numpy as np
+from simtk import openmm, unit
+
+from openmmtools import states, utils
+from openmmtools.constants import ONE_4PI_EPS0
 
-import logging
 logger = logging.getLogger(__name__)
 
-#=============================================================================================
-# PARAMETERS
-#=============================================================================================
 
-ONE_4PI_EPS0 = 138.935456 # OpenMM constant for Coulomb interactions (openmm/platforms/reference/include/SimTKOpenMMRealType.h) in OpenMM units
-                          # TODO: Replace this with an import from simtk.openmm.constants once these constants are available there
+# =============================================================================
+# ALCHEMICAL STATE
+# =============================================================================
+
+class AlchemicalStateError(states.ComposableStateError):
+    """Error raised by an AlchemicalState."""
+    pass
 
-#=============================================================================================
-# MODULE UTILITIES
-#=============================================================================================
 
-def _is_periodic(system):
+class AlchemicalFunction(object):
```

Comment:
Yep! This is a feature that was already implemented in alchemy. Here I just moved them from the `CustomForce` to `AlchemicalState`.